theano/tensor/nnet/nnet.py

```diff
@@ -77,11 +77,16 @@ def perform(self, node, input_storage, output_storage):
         if b.shape[0] != x.shape[1]:
             raise ValueError('b must have same number of columns as x')
 
-        sm = numpy.zeros_like(x)
-        for i in xrange(sm.shape[0]):
-            row = x[i] + b
-            sm[i] = numpy.exp(row - numpy.max(row))
-            sm[i] *= 1.0 / numpy.sum(sm[i])
+        # sm = numpy.zeros_like(x)
+        # for i in xrange(sm.shape[0]):
+            # row = x[i] + b
+            # sm[i] = numpy.exp(row - numpy.max(row))
+            # sm[i] *= 1.0 / numpy.sum(sm[i])
+        # output_storage[0][0] = sm
+
+        x_plus_b = x + b[None, :]
+        e_x = numpy.exp(x_plus_b - x_plus_b.max(axis=1)[:, None])
+        sm = e_x / e_x.sum(axis=1)[:, None]
```

Comment:
This is good change, but can you keep the way of doing with 1.0 / sum(), this was faster as division is much slower the multiplication.

sm = (e_x * (1.0 / e_x.sum(axis=1))[:, None]